rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1082 @@
+            min_value: ScalarValue::Utf8(None),
+            max_value: ScalarValue::Utf8(None),
+        };
+    }
+
+    if max_value.len() > BINARY_PREFIX_LENGTH {
+        max_value = truncate_max_value(&max_value);
+    }
+    if min_value.len() > BINARY_PREFIX_LENGTH {
+        min_value = truncate_min_value(&min_value);
+    }
+    let min_value = str::from_utf8(&min_value).unwrap();
+    let max_value = str::from_utf8(&max_value).unwrap();
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::Utf8(Some(min_value.to_string())),
+        max_value: ScalarValue::Utf8(Some(max_value.to_string())),
+    }
+}
+
+fn get_binary_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut min_value = vec![u8::MAX; BINARY_PREFIX_LENGTH];
+    let mut max_value = vec![u8::MIN; BINARY_PREFIX_LENGTH];
+    let mut null_count: i64 = 0;
+
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<BinaryArray>().unwrap())
+        .collect::<Vec<_>>();
+
+    for array in arr.iter() {
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                if let Some(Ordering::Greater) = value.partial_cmp(max_value.as_slice()) {
+                    max_value = value.to_vec();
+                }
+                if let Some(Ordering::Less) = value.partial_cmp(min_value.as_slice()) {
+                    min_value = value.to_vec();
+                }
+            };
+        });
+        null_count += array.null_count() as i64;
+    }
+    let min_value_scalar = if min_value.len() > BINARY_PREFIX_LENGTH {
+        None
+    } else {
+        Some(min_value)
+    };
+    let max_value_scalar = if max_value.len() > BINARY_PREFIX_LENGTH {
+        None
+    } else {
+        Some(max_value)
+    };
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::Binary(min_value_scalar),
+        max_value: ScalarValue::Binary(max_value_scalar),
+    }
+}
+
+fn get_boolean_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let mut true_present = false;
+    let mut false_present = false;
+
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<BooleanArray>().unwrap())
+        .collect::<Vec<_>>();
+
+    for array in arr.iter() {
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                if value {
+                    true_present = true;
+                } else {
+                    false_present = true;
+                }
+            };
+        });
+        if true_present && false_present {
+            break;
+        }
+    }
+
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(0)),
+        min_value: if false_present {
+            ScalarValue::Boolean(Some(false))
+        } else {
+            ScalarValue::Boolean(None)
+        },
```

Comment:
This isn't quite what we want. If it's all `true`, then we want min and max to be `true`.  Similar for if it's all `false`: we want min and max to be `false`. 

```suggestion
        min_value: if false_present {
            ScalarValue::Boolean(Some(false))
        } else if true_present {
            ScalarValue::Boolean(Some(true))
        } else {
            ScalarValue::Boolean(None)
        },
```